Iterate mem_children, children, io_children

david-macmahon opened the pull request:
Change the iteration order to make it match with `print_topology` and `lstopo` iteration order.  In addition to improved consistency, this is also motivated by the desire to have NUMANode objects be seen before other siblings and their descendants.

david-macmahon committed:
Iterate mem_children, children, io_children

Change the iteration order to make it match with `print_topology` and
`lstopo` iteration order.  In addition to improved consistency, this is
also motivated by the desire to have NUMANode objects be seen before
other siblings and their descendants.

JBlaschke commented on src/lowlevel_api.jl:
```diff
@@ -239,17 +239,17 @@ IteratorEltype(::Type{Object}) = Base.HasEltype()
     # depth-first traversal
     # obj = shift!(state)
     obj, state = state[1], state[2:end]
+    prepend!(state, obj.io_children)
```
Note that this order: `io_children -> children -> memory_children`, does not match the order above: `memory_children -> children -> io_children`. If this memory-children order is preferred then this should be:
```julia
prepend!(state, obj.memory_children)
prepend!(state, obj.children)
prepend!(state, obj.io_children)
```

david-macmahon replied:
Actually, these are `prepend!` so it's "last prepended == first iterated"

david-macmahon replied:
Hmm, maybe it should be `prepend!(state, reverse(obj.memory_children))` etc.

JBlaschke replied:
Ah! this makes sense. I'm not thinking if we should switch to `push!` instead for code clarity? I inherited `prepend!` from @carstenbauer -- maybe you have thoughts.

david-macmahon replied:
We need to use `prepend!` (rather than `push!`) here to keep things iterating in the proper order.  Upon further reflection, we do _not_ need to use `reverse` as I had been considering in an earlier comment.  I think adding a comment to these `prepend!` lines would be helpful to future readers so they don't have to "rediscover" the logic behind this.  I'll push another commit that adds a comment to these lines.

JBlaschke commented:
Also just ran this on Perlmutter -- it seems to work fine. 

I don't know why this is necessary though. @david-macmahon can you explain why we want to iterate over the `memory_children` first? Also note my review indicating a possible typo.

david-macmahon commented:
There are two reasons to iterate `memory_children` first:
1. It makes the (PreOrderDFS) iteration order match the ordering of `print_topology` and `lstopo`
2. It means that `NUMANode` objects will be encountered before the objects that are associated with that `NUMANode`.  This, for example, facilitates matching GPUs to their NUMANode.

I responded to your possible typo comment inline above.  Thanks!

JBlaschke commented:
> There are two reasons to iterate `memory_children` first:
> 
> 1. It makes the (PreOrderDFS) iteration order match the ordering of `print_topology` and `lstopo`
> 2. It means that `NUMANode` objects will be encountered before the objects that are associated with that `NUMANode`.  This, for example, facilitates matching GPUs to their NUMANode.
> 
> I responded to your possible typo comment inline above. Thanks!

Thanks for humoring me. All this looks good from my end. If the `prepend!` discussion is drifting too far into pedantry, just say the word and we can merge now (and deal with questions of code clarity later)

david-macmahon committed:
Add comment about iteration ordering

The comment explains why the children groups are prepended to `state` in
the reverse order of they are iterated.  This is a friendly reminder to
readers (and the author's future self) that this is actually as intended
and not a bug.
